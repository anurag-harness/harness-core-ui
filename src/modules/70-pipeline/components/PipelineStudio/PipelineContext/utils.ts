import { EntityGitDetails } from 'services/cd-ng'

export const getId = (
  accountIdentifier: string,
  orgIdentifier: string,
  projectIdentifier: string,
  pipelineIdentifier: string,
  repoIdentifier = '',
  branch = ''
): string =>
  `${accountIdentifier}_${orgIdentifier}_${projectIdentifier}_${pipelineIdentifier}_${repoIdentifier}_${branch}`

export const getRepoIdentifierName = (gitDetails?: EntityGitDetails): string => {
  return gitDetails?.repoIdentifier || gitDetails?.repoName || ''
}
