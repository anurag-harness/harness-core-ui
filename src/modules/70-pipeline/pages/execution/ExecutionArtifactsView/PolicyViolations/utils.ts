import { queryParamDecodeAll, useQueryParamsOptions } from '@common/hooks/useQueryParams'

export type TableQueryParams = {
  searchTerm?: string
  page?: number
  size?: number
}

export const ENFORCEMENT_VIOLATIONS_PAGE_SIZE = 20
export const ENFORCEMENT_VIOLATIONS_PAGE_INDEX = 0
export const ENFORCEMENT_VIOLATIONS_DEFAULT_SORT = 'name'
export const ENFORCEMENT_VIOLATIONS_DEFAULT_SORT_ORDER = 'ASC'

export const useEnforcementViolationsQueryParamOptions = () => {
  return useQueryParamsOptions({
    page: ENFORCEMENT_VIOLATIONS_PAGE_INDEX,
    size: ENFORCEMENT_VIOLATIONS_PAGE_SIZE,
    sort: ENFORCEMENT_VIOLATIONS_DEFAULT_SORT,
    order: ENFORCEMENT_VIOLATIONS_DEFAULT_SORT_ORDER,
    searchTerm: ''
  })
}

export type PageOptions = {
  page: number
  size: number
  sort: string
  order: string
  searchTerm: string | undefined
}

export const getQueryParamOptions = () => ({
  parseArrays: true,
  decoder: queryParamDecodeAll(),
  processQueryParams(params: PageOptions): PageOptions {
    return {
      ...params,
      page: params.page ?? ENFORCEMENT_VIOLATIONS_PAGE_INDEX,
      size: params.size ?? ENFORCEMENT_VIOLATIONS_PAGE_SIZE,
      sort: params.sort ?? ENFORCEMENT_VIOLATIONS_DEFAULT_SORT,
      order: params.order ?? ENFORCEMENT_VIOLATIONS_DEFAULT_SORT_ORDER
    }
  }
})
