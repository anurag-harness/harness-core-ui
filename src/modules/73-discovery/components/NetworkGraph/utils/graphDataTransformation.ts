/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { Node, Edge } from 'reactflow'
import { NodeTypes, nodeGroupOptions, nodeOptions } from '@discovery/components/NetworkGraph/constants'
import type {
  ApiCreateNetworkMapRequest,
  ApiListDiscoveredServiceConnection,
  ApiListDiscoveredService
} from 'services/servicediscovery'
import type {
  NetworkMapEdgeData,
  NetworkMapNodeData,
  ServiceEdgeData,
  ServiceNodeData
} from '@discovery/components/NetworkGraph/types'

export function getGraphNodesFromServiceList(
  serviceList: ApiListDiscoveredService | null
): Node<ServiceNodeData, NodeTypes>[] {
  if (!serviceList) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<ServiceNodeData, NodeTypes>[] = []
  serviceList.items?.map(service => {
    const namespace = service.spec.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        id: service.id,
        data: { ...service, name: service.spec.kubernetes?.name ?? 'service' },
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromServiceConnections(
  connectionList: ApiListDiscoveredServiceConnection | null
): Edge<ServiceEdgeData>[] {
  if (!connectionList) return []

  const graphEdges: Edge<ServiceEdgeData>[] = []

  connectionList.items?.map(connection => {
    if (connection.id && connection.sourceID && connection.destinationID)
      graphEdges.push({
        id: connection.id,
        source: connection.sourceID,
        target: connection.destinationID,
        data: { parentNode: connection.sourceNamespace, ...connection }
      })
  })

  return graphEdges
}

export function getGraphNodesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Node<NetworkMapNodeData, NodeTypes>[] {
  if (!networkMap) return []

  const namespaces = new Set<string>()
  const graphNodes: Node<NetworkMapNodeData, NodeTypes>[] = []

  networkMap.resources?.map(service => {
    const namespace = service.kubernetes?.namespace
    if (namespace && !namespaces.has(namespace)) {
      namespaces.add(namespace)
    }

    if (service.id)
      graphNodes.push({
        ...nodeOptions,
        type: NodeTypes.NetworkMapHexagon,
        id: service.id,
        data: service,
        parentNode: namespace
      })
  })

  namespaces.forEach(value => {
    graphNodes.push({
      ...nodeGroupOptions,
      id: value,
      data: { id: value, name: value, kind: 'discoveredservice' }
    })
  })

  return graphNodes
}

export function getGraphEdgesFromNetworkMap(
  networkMap: ApiCreateNetworkMapRequest | undefined
): Edge<NetworkMapEdgeData>[] {
  if (!networkMap) return []

  const graphEdges: Edge<NetworkMapEdgeData>[] = []

  networkMap.connections?.map(connection => {
    if (connection.from?.id && connection.to?.id)
      graphEdges.push({
        id: `${connection.from.id}-${connection.to.id}`,
        source: connection.from.id,
        target: connection.to.id,
        data: { ...connection, parentNode: connection.from.kubernetes?.namespace }
      })
  })

  return graphEdges
}
