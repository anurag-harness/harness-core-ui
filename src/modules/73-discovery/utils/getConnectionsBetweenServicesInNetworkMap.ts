/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import {
  ApiListDiscoveredServiceConnection,
  DatabaseConnection,
  DatabaseConnectionType,
  DatabaseNetworkMapEntity
} from 'services/servicediscovery'

export default function getConnectionsBetweenServicesInNetworkMap(
  services: DatabaseNetworkMapEntity[] | null | undefined,
  connectionList: ApiListDiscoveredServiceConnection | null
): DatabaseConnection[] | undefined {
  // Checks for new connections between the selected services
  const serviceHashMap = new Map<string, DatabaseNetworkMapEntity>(services?.map(s => [s.id ?? '', s]))

  const connectionBetweenSelectedServices: DatabaseConnection[] | undefined = connectionList?.items
    ?.filter(conn => serviceHashMap.has(conn.sourceID ?? '') && serviceHashMap.has(conn.destinationID ?? ''))
    .map(conn => ({
      from: {
        id: conn.sourceID,
        kind: serviceHashMap.get(conn.sourceID ?? '')?.kind ?? 'discoveredservice',
        name: conn.sourceName,
        kubernetes: {
          namespace: conn.sourceNamespace
        }
      },
      port: conn.destinationPort,
      to: {
        id: conn.destinationID,
        kind: serviceHashMap.get(conn.destinationID ?? '')?.kind ?? 'discoveredservice',
        name: conn.destinationName,
        kubernetes: {
          namespace: conn.destinationNamespace
        }
      },
      type: conn.type as DatabaseConnectionType,
      manual: false
    }))

  return connectionBetweenSelectedServices
}
