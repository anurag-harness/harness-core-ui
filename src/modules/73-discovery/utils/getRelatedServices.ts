/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import {
  ApiListDiscoveredServiceConnection,
  ApiListDiscoveredService,
  DatabaseDiscoveredServiceCollection
} from 'services/servicediscovery'

export function getRelatedServices(
  serviceID: string,
  serviceList: ApiListDiscoveredService | null,
  connectionList: ApiListDiscoveredServiceConnection | null
): DatabaseDiscoveredServiceCollection[] | undefined {
  if (!serviceList || !connectionList) return

  const relatedServiceIds = connectionList.items?.map(conn => {
    if (serviceID === conn.sourceID) return conn.destinationID ?? ''
    /* istanbul ignore next */
    if (serviceID === conn.destinationID) return conn.sourceID ?? ''
  })

  return serviceList?.items?.filter(service => service.id === serviceID || relatedServiceIds?.includes(service.id))
}
