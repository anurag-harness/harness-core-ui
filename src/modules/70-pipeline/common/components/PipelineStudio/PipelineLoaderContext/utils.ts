import { DefaultNewPipelineId, RouteState } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineActions'
import { EntityGitDetails } from 'services/cd-ng'
import { getId, getRepoIdentifierName } from '@pipeline/components/PipelineStudio/PipelineContext/utils'
import { IDB } from '@modules/10-common/components/IDBContext/IDBContext'

export const deletePipelineFromIDB = async <T = unknown>({
  routeState,
  identifier,
  gitDetails,
  idb
}: {
  routeState: RouteState
  identifier: string
  gitDetails?: EntityGitDetails
  idb: IDB<T>
}): Promise<void> => {
  const id = getId(
    routeState.accountIdentifier,
    routeState.orgIdentifier || '',
    routeState.projectIdentifier || '',
    identifier,
    getRepoIdentifierName(gitDetails),
    gitDetails?.branch || ''
  )
  await idb.del(id)

  const defaultId = getId(
    routeState.accountIdentifier,
    routeState.orgIdentifier || '',
    routeState.projectIdentifier || '',
    DefaultNewPipelineId,
    getRepoIdentifierName(gitDetails),
    gitDetails?.branch || ''
  )
  await idb.del(defaultId)
}
