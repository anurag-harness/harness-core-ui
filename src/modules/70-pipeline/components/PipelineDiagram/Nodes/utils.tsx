/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { isNumber } from 'lodash-es'
import { yamlStringify } from '@common/utils/YamlHelperMethods'
import { dragPlaceholderImageBase64 } from './assets/dragImageBase64'
import { dragStagePlaceholderImageBase64 } from './assets/dragStageImageBase64'
import defaultCss from './DefaultNode/DefaultNode.module.scss'

interface DotNotationPathProps {
  identifier: string
  entityType?: string
  baseDotNotation?: string
}

export enum NodeEntity {
  STAGE = 'STAGE',
  STEP = 'STEP'
}
export enum NodeWrapperEntity {
  step = 'step',
  stepGroup = 'stepGroup',
  stage = 'stage'
}

export const COLLAPSED_MATRIX_NODE_LENGTH = 8
export const MAX_ALLOWED_MATRIX_COLLAPSED_NODES = 4
export const DEFAULT_MATRIX_PARALLELISM = 1

export const getPositionOfAddIcon = (props: any, isRightNode?: boolean): string => {
  if (isRightNode) {
    return '-40px'
  }
  if (props?.children?.length) {
    if (props?.prevNode?.children?.length) {
      return '-65px'
    }
    if (props?.prevNode) return '-58px'
  }
  if (props?.prevNode?.children?.length) {
    return '-58px'
  }
  if (props?.parentIdentifier && !props.prevNode) {
    return '-35px'
  }
  return '-50px'
}

export const transformMatrixLabels = (nodeData: string): string => {
  const parsedNodeName = {}
  try {
    // if object parse to yaml
    const parsedVal = JSON.parse(nodeData)
    if (parsedVal && !isNumber(parsedVal)) {
      Object.assign(parsedNodeName, parsedVal)
      return yamlStringify(parsedNodeName, { indent: 4 })
    } else {
      return JSON.stringify(nodeData)
    }
  } catch (_e) {
    // name is string/number, parse it to string
    return JSON.stringify(nodeData)
  }
}

export const attachDragImageToEventHandler = (event: React.DragEvent<HTMLDivElement>, type?: NodeEntity): void => {
  // set drag image preview to custom icon in case of safari browser, as safari blocks image preview if dom tree have transform property on any parent
  if (navigator.userAgent.search('Safari') >= 0 && navigator.userAgent.search('Chrome') < 0) {
    const dragImage = document.createElement('img')
    dragImage.src = type === NodeEntity.STAGE ? dragStagePlaceholderImageBase64 : dragPlaceholderImageBase64
    dragImage.style.position = 'absolute'
    dragImage.style.left = '-100%'

    document.body.append(dragImage)
    setTimeout(() => dragImage.remove(), 0)

    event.dataTransfer.setDragImage(dragImage, dragImage.width / 2, dragImage.height / 2)
  }
}

export const getEntityIdentifierBasedDotNotationPath = ({
  baseDotNotation,
  identifier,
  entityType = NodeWrapperEntity.step
}: DotNotationPathProps): string => {
  return `${baseDotNotation}.${entityType}.${identifier}`
}

export const getBaseDotNotationWithoutEntityIdentifier = (dotNotation = ''): string => {
  const lastDotIndex = dotNotation.lastIndexOf('.')
  return lastDotIndex !== -1 ? dotNotation.substring(0, lastDotIndex) : dotNotation
}

export const getConditionalClassName = (isDisabled: boolean, className?: string): { className?: string } => {
  return isDisabled ? { className: defaultCss.disabledIcon } : className ? { className } : {}
}
