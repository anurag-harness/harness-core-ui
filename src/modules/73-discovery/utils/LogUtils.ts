/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { sanitizeHTML } from '@common/utils/StringUtils'
import { formatDatetoLocale } from '@common/utils/dateUtils'
import { loggerFor } from 'framework/logging/logging'
import { ModuleName } from 'framework/types/ModuleName'
import { DatabaseInstallationCollection } from 'services/servicediscovery'

export interface ArgsType {
  PodName: string
  PodNamespace: string
}

export interface DAgentLogs {
  clusterLogs: string
  nodeLogs: string
}

/**
 * Data for a log unit/section
 */
export interface LogSectionData {
  /**
   * The log data
   */
  data: LogLineData[]
}

/**
 * Keys in the json from server
 */
export type TextKeys = 'level' | 'out' | 'time' | 'args'

export interface LogLineData {
  /**
   * Text for the line
   */
  text: Partial<Record<TextKeys, string | ArgsType>>
  /**
   * Indices of the search result with in the current line.
   * This will be a continous list
   *
   * If this values is `[3, 4]`, it means the result number 3 and 4 will be found in this line
   */
  searchIndices?: Partial<Record<TextKeys, number[]>>
}

const logger = loggerFor(ModuleName.CHAOS)

export function processLogsData(data: string): LogLineData[] {
  return String(data)
    .split('\n')
    .reduce<LogSectionData['data']>((accumulator, line) => {
      /* istanbul ignore else */
      if (line.length > 0) {
        try {
          const { level, time, out, args } = JSON.parse(line) as Record<string, string>
          accumulator.push({
            text: {
              level: sanitizeHTML(level),
              time: formatDatetoLocale(time),
              out: sanitizeHTML(out),
              args: args
            }
          })
        } catch (error) {
          logger.error(`Error while parsing logs: ${error}`)
        }
      }
      return accumulator
    }, [])
}

export function processAgentLogs(
  logData: string | undefined,
  installationList: DatabaseInstallationCollection | undefined
): DAgentLogs {
  const logsData = processLogsData(logData ?? '')
  let clusterLogs = ''
  let nodeLogs = ''
  const nodeNames: Array<string> = []
  installationList?.agentDetails?.node?.forEach(node => {
    if (node?.name) nodeNames.push(node?.name)
  })
  logsData.map(log => {
    if (installationList?.agentDetails?.cluster?.name === (log.text.args as ArgsType)?.PodName) {
      clusterLogs = clusterLogs + log.text.out + '\n'
    }
    if (nodeNames.includes((log.text.args as ArgsType)?.PodName)) {
      nodeLogs = nodeLogs + log.text.out + '\n'
    }
  })

  return {
    clusterLogs: clusterLogs.length ? clusterLogs : 'No cluster level logs found',
    nodeLogs: nodeLogs.length ? nodeLogs : 'No node level logs found'
  }
}
