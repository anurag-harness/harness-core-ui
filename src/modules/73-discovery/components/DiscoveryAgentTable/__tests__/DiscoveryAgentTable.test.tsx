/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { PopoverPosition } from '@blueprintjs/core'
import { render } from '@testing-library/react'
import type { ApiGetAgentResponse } from 'services/servicediscovery'
import routes from '@common/RouteDefinitions'
import { accountPathProps, projectPathProps } from '@common/utils/routeUtils'
import { TestWrapper } from '@common/utils/testUtils'
import DiscoveryAgentTable from '../DiscoveryAgentTable'

const mockDiscoveryAgent: ApiGetAgentResponse[] = [
  {
    id: '648ffda242d20c0d04713e05',
    name: 'local-1',
    identity: 'local-1',
    description: 'some description',
    accountIdentifier: 'VgWXxi_6TdqAyplTQMg4CQ',
    organizationIdentifier: 'default',
    projectIdentifier: 'test',
    createdAt: '2023-06-21T13:17:41.491Z',
    updatedAt: '2023-06-21T13:17:41.491Z',
    createdBy: 'VgWXxi_6TdqAyplTQMg4CQ',
    updatedBy: 'Sahil',
    installationDetails: {
      createdAt: '2023-06-21T13:17:41.491Z',
      updatedAt: '2023-06-21T13:17:41.491Z',
      createdBy: 'VgWXxi_6TdqAyplTQMg4CQ',
      updatedBy: 'Sahil'
    },
    config: {
      kubernetes: {
        namespaced: false,
        namespace: 'sd1',
        serviceAccount: 'cluster-admin-1',
        imageRegistry: 'index.docker.io/shovan1995',
        imageTag: 'ci',
        imagePullPolicy: 'Always'
      },
      data: {
        enableNodeAgent: false,
        enableStorageResources: false,
        enableBatchResources: false,
        retryInSecond: 100,
        retryCount: 3,
        batchSize: 100,
        collectionWindowInMin: 2
      }
    }
  }
]

const mockPagination = {
  itemCount: 1,
  pageSize: 1,
  pageCount: 1,
  pageIndex: 0,
  gotoPage: jest.fn,
  onPageSizeChange: jest.fn,
  showPagination: true,
  pageSizeDropdownProps: {
    usePortal: true,
    popoverProps: {
      position: PopoverPosition.TOP
    }
  },
  pageSizeOptions: [10, 20, 50, 100]
}

const PATH = routes.toDiscovery({ ...accountPathProps, ...projectPathProps })
const PATH_PARAMS = {
  accountId: 'accountId',
  orgIdentifier: 'default',
  projectIdentifier: 'Discovery_Test',
  module: 'chaos'
}

describe('Discovery Agent Status ', () => {
  beforeAll(() => {
    jest.useFakeTimers().setSystemTime(new Date('2023-06-28'))
  })
  test('render', async () => {
    const props = {
      listData: mockDiscoveryAgent,
      pagination: mockPagination,
      refetch: () => Promise.resolve()
    }
    const { container } = render(
      <TestWrapper path={PATH} pathParams={PATH_PARAMS}>
        <DiscoveryAgentTable {...props} />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
})
