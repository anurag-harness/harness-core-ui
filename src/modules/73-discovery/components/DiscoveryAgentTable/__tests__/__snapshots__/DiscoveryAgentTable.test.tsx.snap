// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discovery Agent Status  render 1`] = `
<div>
  <div
    class="TableV2--table"
  >
    <div
      class="TableV2--header"
      role="row"
    >
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 20%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Discovery Agent
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 20%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Last Discovery
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 15%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Discovery Schedule
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 15%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Services Discovered
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 15%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Network Maps
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 15%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Last Edited
        </p>
      </div>
      <div
        class="TableV2--cell TableV2--sortable"
        colspan="1"
        role="columnheader"
        style="width: 5%;"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-table-headers"
        >
          Â 
        </p>
      </div>
    </div>
    <div
      class="TableV2--body"
    >
      <div
        class="TableV2--row TableV2--card"
        role="row"
      >
        <div
          class="TableV2--cells"
        >
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 20%;"
          >
            <a
              href="/account/accountId/home/orgs/default/projects/Discovery_Test/setup/resources/discovery/local-1"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-normal StyledProps--font-weight-bold StyledProps--color StyledProps--color-primary7"
                style="cursor: pointer;"
              >
                local-1
              </p>
            </a>
            <div
              class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-alignItems-center StyledProps--flex-justifyContent-start StyledProps--margin-top-xsmall"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--font-weight-light StyledProps--color StyledProps--color-grey500"
              >
                common.ID
                :
              </p>
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main idPill StyledProps--font-size-small StyledProps--font-weight-light"
                style="background: rgb(243, 243, 250); border-radius: 5px; --text-line-clamp: 1;"
              >
                local-1
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 20%;"
          >
            <div
              class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-justifyContent-flex-start"
            >
              <div
                class="Layout--vertical Layout--layout-spacing-small StyledProps--main"
              >
                <p
                  class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey900"
                  style="--text-line-clamp: 1;"
                >
                  Jun 19, 2023 07:02 AM
                </p>
                <div
                  class="StyledProps--font StyledProps--main StyledProps--background-green50"
                  style="padding: 2px 6px; border-radius: 4px;"
                >
                  <p
                    class="StyledProps--font StyledProps--main StyledProps--font-size-xsmall StyledProps--color StyledProps--color-green700"
                    style="font-weight: 600;"
                  >
                    SUCCESS
                  </p>
                </div>
              </div>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 15%;"
          >
            <div>
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-small StyledProps--color StyledProps--color-grey400"
              >
                pipeline.notAvailable
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 15%;"
          >
            <div
              class="Layout--vertical StyledProps--main totalServiceContainer"
              style="width: 60px; height: 50px;"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-medium StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey500"
              >
                --
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 15%;"
          >
            <div
              class="Layout--vertical StyledProps--main totalServiceContainer"
              style="width: 50px; height: 40px;"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-size-medium StyledProps--font-weight-semi-bold StyledProps--color StyledProps--color-grey500"
              >
                --
              </p>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 15%;"
          >
            <div
              class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-justifyContent-flex-start"
            >
              <p
                class="StyledProps--font Text--lineclamp StyledProps--main StyledProps--font-size-xsmall StyledProps--color StyledProps--color-grey500"
                style="--text-line-clamp: 1;"
              >
                6 days ago
              </p>
              <div
                class="Avatar--Avatar Avatar--contentStyle"
              >
                <span
                  class="bp3-popover-wrapper Avatar--avatarPopOver"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <div
                      class="Avatar--AvatarInner"
                      style="border-radius: 100%; line-height: 26px; width: 32px; height: 32px; font-size: 10px;"
                      tabindex="0"
                    >
                      S
                    </div>
                  </span>
                </span>
              </div>
            </div>
          </div>
          <div
            class="TableV2--cell"
            role="cell"
            style="width: 5%;"
          >
            <div
              class="Layout--horizontal StyledProps--main"
              style="justify-content: flex-end; margin-right: 10px;"
            >
              <span
                class="bp3-popover-wrapper bp3-dark"
              >
                <span
                  class="bp3-popover-target"
                >
                  <button
                    class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
                    type="button"
                  >
                    <span
                      class="bp3-icon StyledProps--main"
                      data-icon="Options"
                    >
                      <svg
                        fill="currentColor"
                        height="16"
                        viewBox="0 0 4 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <circle
                          cx="2"
                          cy="2"
                          r="1.5"
                        />
                        <circle
                          cx="2"
                          cy="8"
                          r="1.5"
                        />
                        <circle
                          cx="2"
                          cy="14"
                          r="1.5"
                        />
                      </svg>
                    </span>
                  </button>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="Layout--horizontal StyledProps--main TableV2--pagination Pagination--container StyledProps--padding-top-medium StyledProps--padding-bottom-medium StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-distribution-space-between"
    >
      <span
        class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--font-variation-small-semi"
      >
        1
         of 
        1
      </span>
      <div
        class="Layout--horizontal StyledProps--main"
      >
        <button
          aria-label="Prev"
          class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--buttonLeft Button--with-current-color Button--withLeftIcon Button--size Button--small"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="bp3-icon bp3-icon-arrow-left StyledProps--main StyledProps--padding-right-xsmall"
            icon="arrow-left"
          >
            <svg
              data-icon="arrow-left"
              height="12"
              viewBox="0 0 16 16"
              width="12"
            >
              <desc>
                arrow-left
              </desc>
              <path
                d="M13.99 6.99H4.41L7.7 3.7a1.003 1.003 0 00-1.42-1.42l-5 5a1.014 1.014 0 000 1.42l5 5a1.003 1.003 0 001.42-1.42L4.41 8.99H14c.55 0 1-.45 1-1s-.46-1-1.01-1z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
          <span
            class="bp3-button-text"
          >
            Prev
          </span>
        </button>
        <button
          class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--selected Button--with-current-color"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            1
          </span>
        </button>
        <button
          aria-label="Next"
          class="bp3-button bp3-disabled Button--button StyledProps--font StyledProps--main Pagination--roundedButton Pagination--buttonRight Button--with-current-color Button--withRightIcon Button--size Button--small"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="bp3-button-text"
          >
            Next
          </span>
          <span
            class="bp3-icon bp3-icon-arrow-right StyledProps--main"
            icon="arrow-right"
          >
            <svg
              data-icon="arrow-right"
              height="12"
              viewBox="0 0 16 16"
              width="12"
            >
              <desc>
                arrow-right
              </desc>
              <path
                d="M14.7 7.29l-5-5a.965.965 0 00-.71-.3 1.003 1.003 0 00-.71 1.71l3.29 3.29H1.99c-.55 0-1 .45-1 1s.45 1 1 1h9.59l-3.29 3.29a1.003 1.003 0 001.42 1.42l5-5c.18-.18.29-.43.29-.71s-.12-.52-.3-.7z"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
      </div>
      <div
        class="Layout--horizontal Layout--layout-spacing-small StyledProps--main StyledProps--flex StyledProps--flex-alignItems-baseline"
      >
        <p
          class="StyledProps--font StyledProps--main"
        >
          Show
        </p>
        <div
          class="bp3-popover-wrapper DropDown--main Pagination--pageSizeDropdown"
        >
          <div
            class="bp3-popover-target"
          >
            <div
              class=""
            >
              <div
                class="Layout--horizontal StyledProps--main DropDown--dropdownButton DropDown--withBorder StyledProps--flex"
                data-testid="dropdown-button"
                style="width: 60px;"
              >
                <div
                  class="Layout--horizontal StyledProps--main DropDown--labelWrapper StyledProps--flex"
                >
                  <p
                    class="StyledProps--font Text--lineclamp StyledProps--main DropDown--label"
                    data-testid="dropdown-value"
                    style="--text-line-clamp: 1;"
                  >
                    Select
                  </p>
                </div>
                <div
                  class="Layout--horizontal StyledProps--main DropDown--btnWrapper StyledProps--flex"
                >
                  <span
                    class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey400"
                    data-icon="main-chevron-down"
                  >
                    <svg
                      height="8"
                      viewBox="0 0 14 8"
                      width="8"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M13 0c-.3 0-.5.1-.7.3L7 5.6 1.7.3C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7l6 6c.2.2.4.3.7.3s.5-.1.7-.3l6-6c.2-.2.3-.4.3-.7 0-.6-.4-1-1-1z"
                        fill="#4d5969"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <p
          class="StyledProps--font StyledProps--main"
        >
          per page
        </p>
      </div>
    </div>
  </div>
</div>
`;
